torch/_torch_docs.py

```diff
@@ -8011,6 +8011,49 @@ def merge_dicts(*dicts):
             [-0.0881,  0.4370,  0.2275,  1.0284]])
 """.format(**common_args))
 
+add_docstr(torch.sparse_csr_tensor,
+           r"""
+sparse_csr_tensor(crow_indices, col_indices, values, size=None, *, dtype=None, device=None, requires_grad=False) -> Tensor
+
+Constructs a :ref:`sparse tensor in CSR (Compressed Sparse Row) <sparse-csr-docs>` with specified
+values at the given :attr:`crow_indices` and :attr:`col_indices`. Sparse matrix multiplication operations
+in CSR format are typically faster than that for sparse tensors in COO format. Make you have a look
+at :ref:`the note on the data type of the indices <sparse-csr-docs>`.
+
+Args:
+    crow_indices (array_like): One-dimensional array of size size[0] + 1. The last element
+        is the number of non-zeros. This tensor encodes the index in values and col_indices
+        depending on where the given row starts. Each successive number in the tensor
+        subtracted by the number before it denotes the number of elements in a given row.
+    col_indices (array_like): Column co-ordinates of each element in values. Strictly one
+        dimensional tensor with the same length as values.
+    values (array_list): Initial values for the tensor. Can be a list, tuple, NumPy ``ndarray``, scalar,
+        and other types.
+    size (list, tuple, :class:`torch.Size`, optional): Size of the sparse tensor. If not provided, the
+        size will be inferred as the minimum size big enough to hold all non-zero elements.
+
+Keyword args:
+    dtype (:class:`torch.dtype`, optional): the desired data type of returned tensor.
+        Default: if None, infers data type from :attr:`values`.
+    device (:class:`torch.device`, optional): the desired device of returned tensor.
+        Default: if None, uses the current device for the default tensor type
+        (see :func:`torch.set_default_tensor_type`). :attr:`device` will be the CPU
+        for CPU tensor types and the current CUDA device for CUDA tensor types.
```

Comment:
Can you check if we have other API precedent besides COO tensors for what happens for tensor constructors that (1) take tensor arguments, (2) are supposed to be O(1) if they get a tensor, and (3) also take optional dtype/device? It's not clear to me that we want to insert an implicit dtype/device conversion in these cases. I see that these parameters make sense if you pass non-tensor arguments but I'm not sure what the correct API behavior for tensor argument is.

Reply:
Re (3): if `input` is a tensor then we have
1. `torch.as_tensor(input, dtype=dtype, device=device)` is equivalent to `input.to(dtype=dtype, device=device, copy=False)`
2. `torch.tensor(input, dtype=dtype, device=device)` is equivalent to `input.to(dtype=dtype, device=device, copy=True)`

For memory usage efficiency, I would propose to follow `torch.as_tensor` convention when constructing COO/CSR tensors:
3. `torch.sparse_coo_tensor(indices, values, dtype=dtype, device=device)` is equivalent to
```
torch.sparse_coo_tensor(indices.to(dtype=torch.int64, device=device, copy=False),
                        values.to(dtype=dtype, device=device, copy=False))
````
where `indices` and `values` are tensors. A similar equivalence can be defined for CSR tensor constructor where the `dtype` of `crow_indices/col_indices` can be `torch.int32` or `torch.int64`.

Reply:
OK, this seems fine. Conditional aliasing behavior is going to be a pain when we add autograd support but we'll get there when we get there.